### ISSUE TITLE
[Core] Apparent task/object leak on certain direct task calls.

### ISSUE THREAD

**Issue by ClarkZinzow** [ray-project/ray#7887]
### What is the problem?

Ray appears to be leaking tasks/objects on certain direct task calls. Namely, it appears that the raylet/node manager is losing track of direct task calls that depend on the same return object of an upstream task (and are possibly sharing a worker lease, need to confirm that.) So far, all instances of leaked tasks appear to correspond to pairs of tasks, sharing the same scheduler key and submitted nearly at the same time, both depending on the same task return object, with the task submitted second tending to be leaked.

From an end-user perspective, this leak manifests in many orphaned objects being left in the object store, none of which are retrievable (`ray.get()` hangs indefinitely.)

#### Version information

- Python 3.7.5
- Ubuntu 19.10
- ray at commit 93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b (will test on most recently built master commit soon)

### Bug Information

Currently unable to create a minimal reproduction, still working on that. @stephanie-wang said that it'd be a good idea to open an issue so this bug isn't only tracked by a Slack thread, and she was pretty confident that there is in fact a leak. Below is a summary of some log and code spelunking.

#### Summary

- Two tasks, `240f6960faf7db13ffffffff0100` and `c83000702c305163ffffffff0100`, depend on the same object `1d539110471eab4dffffffff010000c801000000`, and both appear to have the same scheduling key and both are making worker lease requests at around the same time.
- It appears that the lease request for the second task, `c83000702c305163ffffffff0100`, is never granted.
- It also appears that the raylet loses track of the second task: the task is never transitioned out of `READY`. Namely, in the node manager, [`FinishAssignTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L3016-L3055) and [`FinishAssignedTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L2412-L2456) are never called for the task.
- An object with ID `1d539110471eab4dffffffff010000c801000000` ends up being orphaned in the object store since the node manager thinks that task `c83000702c305163ffffffff0100` still depends on the object after its unpinned; this results in a failed object pull and a failed object reconstruction, which I believe means that the object orphaned in the object store is an error object.
- Not sure if the second task is reusing the first task's worker lease and is somehow making a lease request of its own, or if they aren't sharing a worker lease and the lease request never returns and the raylet loses track of the task for a different reason.

#### Details

There are two downstream tasks that depend on the return object of task `1d539110471eab4dffffffff0100`:

raylet:
```
I0331 01:26:54.120884  4315 task_dependency_manager.cc:184] Task 240f6960faf7db13ffffffff0100 blocked on object 1d539110471eab4dffffffff010000c801000000
I0331 01:26:55.895077  4315 task_dependency_manager.cc:184] Task c83000702c305163ffffffff0100 blocked on object 1d539110471eab4dffffffff010000c801000000
```

Both of these downstream tasks complete according to the worker:

python-worker:
```
I0331 01:26:55.904520  4339 core_worker.cc:1186] Finished executing task 240f6960faf7db13ffffffff0100
I0331 01:26:55.909976  4352 task_manager.cc:127] Completing task 240f6960faf7db13ffffffff0100
I0331 01:26:55.930382  4339 core_worker.cc:1186] Finished executing task c83000702c305163ffffffff0100
I0331 01:26:55.930879  4352 task_manager.cc:127] Completing task c83000702c305163ffffffff0100
```

However, according to the raylet logs, task `240f6960faf7db13ffffffff0100` finished but task `c83000702c305163ffffffff0100` did not:

raylet:
```
I0331 01:26:55.950846  4315 node_manager.cc:2414] Finished task 240f6960faf7db13ffffffff0100
```

In fact, from the raylet's perspective, the `c83000702c305163ffffffff0100`  task is never moved out of the ready queue into the running queue by the scheduler, the node manager never assigns the task to a worker, etc. I'm assuming that this is because that task is a direct task call whose scheduling decision is cached and an existing worker lease (provisioned by task `240f6960faf7db13ffffffff0100`) is being reused, so no worker lease is granted and it appears to go through a different code path that doesn't call [`FinishAssignTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L3016-L3055) or [`FinishAssignedTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L2412-L2456) in the node manager. Therefore, from the node manager's perspective, `c83000702c305163ffffffff0100`  is a queued task that still depends on the return object of task `1d539110471eab4dffffffff0100` when that object gets unpinned:

raylet:
```
I0331 01:26:55.950737  4315 node_manager.cc:3264] Unpinning object 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616161  4315 node_manager.cc:2815] Object missing 1d539110471eab4dffffffff010000c801000000,  on 87b1fb4ac8cb822f02545201693611dc853749e7, 1 tasks waiting, tasks: c83000702c305163ffffffff0100
```

This will cause the task dependency manager to consider the object ID required, which causes an attempted (and failed) object pull and object reconstruction to kick off 10 seconds later:

https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L2804
https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/task_dependency_manager.cc#L170
https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/task_dependency_manager.cc#L73-L74
https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/reconstruction_policy.cc#L210-L223 

raylet:
```
# unpinned
I0331 01:26:55.950737  4315 node_manager.cc:3264] Unpinning object 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616024  4315 object_directory.cc:71] Reporting object removed to GCS 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616030  4315 service_based_accessor.cc:748] Removing object location, object id = 1d539110471eab4dffffffff010000c801000000, node id = 87b1fb4ac8cb822f02545201693611dc853749e7
# attempted object pull request
I0331 01:26:56.616096  4315 object_manager.cc:126] Pull on 87b1fb4ac8cb822f02545201693611dc853749e7 of object 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616108  4315 service_based_accessor.cc:776] Subscribing object location, object id = 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616137  4315 service_based_accessor.cc:781] Finished subscribing object location, object id = 1d539110471eab4dffffffff010000c801000000
# object reconstruction in 10 seconds from this point since pull request will fail
I0331 01:26:56.616145  4315 reconstruction_policy.cc:211] Listening and maybe reconstructing object 1d539110471eab4dffffffff010000c801000000
I0331 01:26:56.616161  4315 node_manager.cc:2815] Object missing 1d539110471eab4dffffffff010000c801000000,  on 87b1fb4ac8cb822f02545201693611dc853749e7, 1 tasks waiting, tasks: c83000702c305163ffffffff0100
# failed object pull
E0331 01:26:56.616714  4315 object_manager.cc:189] The object manager with ID 87b1fb4ac8cb822f02545201693611dc853749e7 is trying to pull object 1d539110471eab4dffffffff010000c801000000 but the object table suggests that this object manager already has the object. The object may have been evicted.
I0331 01:26:56.618919  4315 service_based_accessor.cc:762] Finished removing object location, status = OK, object id = 1d539110471eab4dffffffff010000c801000000, node id = 87b1fb4ac8cb822f02545201693611dc853749e7
# object reconstruction starts here, after waiting for 10 seconds
I0331 01:27:06.616205  4315 service_based_accessor.cc:653] Subscribing task lease, task id = 1d539110471eab4dffffffff0100
I0331 01:27:06.616250  4315 service_based_accessor.cc:658] Finished subscribing task lease, task id = 1d539110471eab4dffffffff0100
I0331 01:27:06.618786  4315 service_based_accessor.cc:574] Getting task, task id = 1d539110471eab4dffffffff0100
# task lineage not available in GCS since it was already cleaned up
W0331 01:27:06.619911  4315 node_manager.cc:2703] Metadata of task 1d539110471eab4dffffffff0100 not found in either GCS or lineage cache. It may have been evicted by the redis LRU configuration. Consider increasing the memory allocation via ray.init(redis_max_memory=<max_memory_bytes>).
I0331 01:27:06.620035  4315 service_based_accessor.cc:585] Finished getting task, status = Invalid: Task not exist., task id = 1d539110471eab4dffffffff0100
```

After the reconstruction fails, it appears that an object is put into the object store under that ID, which I believe is an error object (I can't inspect it since the object has already gone out of scope according to the Ray workers; a `ray.get()` call hangs indefinitely):

```
I0331 01:27:06.620035  4315 service_based_accessor.cc:585] Finished getting task, status = Invalid: Task not exist., task id = 1d539110471eab4dffffffff0100
I0331 01:27:06.620261  4315 object_manager.cc:78] Object added 1d539110471eab4dffffffff010000c801000000
I0331 01:27:06.620273  4315 object_directory.cc:62] Reporting object added to GCS 1d539110471eab4dffffffff010000c801000000
I0331 01:27:06.620281  4315 service_based_accessor.cc:721] Adding object location, object id = 1d539110471eab4dffffffff010000c801000000, node id = 87b1fb4ac8cb822f02545201693611dc853749e7
I0331 01:27:06.620378  4315 service_based_accessor.cc:787] Unsubscribing object location, object id = 1d539110471eab4dffffffff010000c801000000
I0331 01:27:06.620412  4315 service_based_accessor.cc:789] Finished unsubscribing object location, object id = 1d539110471eab4dffffffff010000c801000000
I0331 01:27:06.620429  4315 reconstruction_policy.cc:226] Reconstruction for object 1d539110471eab4dffffffff010000c801000000 canceled
I0331 01:27:06.620440  4315 service_based_accessor.cc:664] Unsubscribing task lease, task id = 1d539110471eab4dffffffff0100
I0331 01:27:06.620460  4315 service_based_accessor.cc:666] Finished unsubscribing task lease, task id = 1d539110471eab4dffffffff0100
I0331 01:27:06.620471  4315 node_manager.cc:2756] Object local 1d539110471eab4dffffffff010000c801000000,  on 87b1fb4ac8cb822f02545201693611dc853749e7, 1 tasks ready
I0331 01:27:06.620859  4315 service_based_accessor.cc:736] Finished adding object location, status = OK, object id = 1d539110471eab4dffffffff010000c801000000, node id = 87b1fb4ac8cb822f02545201693611dc853749e7
```

I believe that the reference counting logic is behaving as expected. It appears to be the `c83000702c305163ffffffff0100` task becoming untracked by the raylet/node manager that's causing the failed object pull and the failed task reconstruction. Whatever code path this task is taking, it needs to tell the raylet/node manager that the task is assigned (once it's assigned) and that it's finished (once it's finished.)

#### Direct Task Model Confusion

I'm still trying to wrap my head around the code path differences between a direct task that leases a worker and a direct task that reuses an existing worker lease. In this case, it appears that both tasks request a worker lease:

raylet:
```
I0331 01:26:54.120754  4315 node_manager.cc:1626] Worker lease request 240f6960faf7db13ffffffff0100
I0331 01:26:55.894937  4315 node_manager.cc:1626] Worker lease request c83000702c305163ffffffff0100
```

but only one is granted:

python-driver
```
I0331 01:26:55.894311  4352 direct_task_transport.cc:133] Lease granted 240f6960faf7db13ffffffff0100
```

suggesting that (if the lease is shared) a lease request is made for each task once dependencies are resolved, the lease is granted for the first task while the second task's lease request is not, the latter task is pushed to the worker after the former completes, but the former never calls the [`task_done_`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/core_worker/transport/direct_actor_transport.cc#L283-L290) callback, which never causes the raylet to indicate that the worker is available, so the worker is never put back into the worker pool, which causes the latter task to never be "assigned" (via [`NodeManager.AssignTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L2327-L2410) or [`NodeManager.FinishAssignTask()`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L3016-L3055)) to a worker, so the task is never transitioned to `RUNNING`.

However, I can't see how two lease requests on the same scheduling key could be in flight concurrently, given that the direct task submitter's [`RequestNewWorkerIfNeeded`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/core_worker/transport/direct_task_transport.cc#L108-L151) is guarded by a mutex and a collision in `pending_lease_requests_` should short-circuit the function, preventing the second task from submitting a lease request. This would suggest that the second task isn't sharing a worker lease, but that makes the fact that the lease is never granted and the raylet losing track of the task even more confusing. Moreover, I've pretty sure that both tasks have the same scheduling key (they have the same resource requirements, function descriptor, and arguments.)

### Next Steps

- If anyone could shed some light on the intended submission flow and lifetime of direct tasks reusing worker leases, I would be forever grateful.
- I'm going to keep working on a minimal reproducible example and will add it to this issue if/when I find one.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://ray.readthedocs.io/en/latest/installation.html).

**Comment by ClarkZinzow** [comment]
@stephanie-wang That's awesome! And that sequence jives with my spelunking, a few quick questions there:

>Driver sends one lease request for each task to the raylet, which queues them. We submit one lease request for each because optimistically, we'd like the tasks to run in parallel on different workers.

I thought that only one lease request should be sent for tasks with the same scheduling key? Shouldn't the T1 lease request be made, with that lease request added to `pending_lease_requests_`, while holding a lock on a lease request mutex, and then when `RequestNewWorkerIfNeeded` is called for T2, it should short-circuit before making a lease request for T2?

https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/core_worker/transport/direct_task_transport.cc#L110-L113

I'm probably missing something here, but it seems like those concurrent lease requests shouldn't be possible. 🤔

>Raylet still has T2's lease request queued and is trying to fetch O since it is no longer available locally. Eventually, this times out because O will not be created again.

So the flow for this is:

1. The object is deleted from the object store.
2. This triggers a `SubscribeObjDeleted` subscription, which [calls `HandleObjectMissing`](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/node_manager.cc#L140-L141).
3. Which calls the corresponding `HandleObjectMissing` method on the dependency manager.
4. Which checks to see if the remote dependency is required; it _is_ considered required since it was [never removed from the set of required tasks](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/raylet/task_dependency_manager.cc#L43).
5. This triggers the immediate failed pull request and the failed reconstruction attempt 10 seconds later.

Is that about right?

>My plan to fix this is to have the worker cancel active lease requests once there are no more tasks to run for that scheduling key. Downside is we'd need to keep a bit more state at the workers, and have an additional IPC to send the cancellation. This could also slightly hurt performance if another task with the same key gets added immediately after, since we'd have to request a lease from the raylet again.

I'm probably missing something, but shouldn't the task dependency manager be notified of a batched task being done right after the task finishes, via the existing callbacks, rather than on lease cancellation at the end of a batch? I could imagine a batch of tasks fitting within a lease window pilling up, none finishing according to the raylet since it won't receive notice that the tasks have finished until those lease request cancellations come through.

Is this an apt description of ideal behavior?
- The raylet properly maintains the task state for batched tasks, transitioning them into `RUNNING` when the task is sent to the worker and removing them from the node manager's `local_queue_` when the task is done executing.
- The task should be removed from the raylet's task dependency manager (via `task_dependency_manager_.UnsubscribeGetDependencies()` and `task_dependency_manager_.Cancel()`) when the task is done running.
- The worker lease expiration is extended at the beginning of the execution of a batched task.

If so, could that be achieved by the node manager's `FinishAssignTask` and `FinishAssignedTask` methods are invoked at the beginning of task execution and after task execution has finished (e.g. in the core worker direct call receiver's `HandlePushTask` method, via a mechanism similar to [the `OnTaskDone` callback](https://github.com/ray-project/ray/blob/93b5c38b7dc75b64a4812c1b300f48f9f1bf5d2b/src/ray/core_worker/transport/direct_actor_transport.cc#L289)), respectively, with the task lease request returning an indication of a successful reuse of an existing lease (indicating that the lease expiration should be extended)? I feel like the semantics might be better when having the lease request return with an extension indication and having the direct task receiver tell the raylet about task state transitions so that the same `FinishAssignTask` and `FinishAssignedTask` methods are used to keep the raylet up-to-date. What do you think?

Also I wouldn't be surprised if I completely misunderstood a billion things about the leasing architecture and if everything that I said is wrong or untenable, feel free to ignore if that's the case! 😁

**Comment by stephanie-wang** [comment]
Hey @ClarkZinzow, if you get a chance, can you try out your application on the latest master? We just merged #7929 which should hopefully have fixed the issue. Thanks!

**Comment by ClarkZinzow** [comment]
Thanks @stephanie-wang, that was quick! 😁 I'll confirm that the issue is fixed today.

**Comment by ClarkZinzow** [comment]
@stephanie-wang I haven't been able to trigger any more task leakages, so it looks like the lease cancellation has done the trick!

**Comment by stephanie-wang** [comment]
Great, thanks @ClarkZinzow for reporting and helping us out with the issue!
